### DISCUSSION TITLE
change jsconfig-paths-plugin to resolve paths synchronously

### DISCUSSION THREAD

**Discussion by jantimon** [vercel/next.js#29421] [Ideas]
### Describe the feature you'd like to request

I am working on a nextjs POC with linaria. ([linaria](https://github.com/callstack/linaria/) is using webpack & babel to extract css from css-in-js at build time)

the `@linaria/webpack-loader` is able to [resolve dependencies](https://github.com/callstack/linaria/blob/c49df269ebbbc956d43720b007a92f9f865662c5/packages/webpack5-loader/src/index.ts#L59-L70) based on the current webpack configuration with `resolveSync` of `enhanced-require`.

unfortunately the `next/packages/build/webpack/plugins/jsconfig-paths-plugin` is [promise based](https://github.com/vercel/next.js/blob/2271cd841d9840b703f57fb8d791b4e3b373430d/packages/next/build/webpack/plugins/jsconfig-paths-plugin.ts#L215) and therefore `enhanced-require` throws [`Cannot 'resolveSync' because the fileSystem is not sync`](https://github.com/webpack/enhanced-resolve/blob/d1a3cb65cdf90754871a263172006e2902a2f86e/lib/Resolver.js#L241-L245)

`linaria` is not able to resolve dependencies asynchronously as `babel` (and also `swc`) allow only synchronous path resolvers

### Describe the solution you'd like

- turning the promise based hook into a callback based hook (I can prepare a PR if that helps)

### Describe alternatives you've considered

- babel and/or swc could allow async path resolving ([babel/babel/issues/13798](https://github.com/babel/babel/issues/13798))

**Comment by sokra** [comment]
Sounds good, send a PR

enhanced-resolve exposes a `forEachBail` helper for the loop which supports sync and async execution: https://github.com/webpack/enhanced-resolve/blob/e5ff68aef5ab43b8197e864181eda3912957c526/lib/index.js#L128

**Reply by jantimon** [reply]
Thanks tobias :)

I created a first PR and it seems to work https://github.com/vercel/next.js/pull/29467
Just let me know if you prefer to use `forEachBail` instead

Also I noticed that `triedPaths` is unused..

**Reply by jantimon** [reply]
I copied over forEachBail from enhanced-resolve as you suggested
